### PR TITLE
parser: implement Restore for TableName

### PR DESCRIPTION
**Pull request by feloxx**
1. Add a string map of the const field

2. Add TableName Restore (There will be a logical call to TableName when it is empty, so there is a judgment empty processing)

3. Add IndexHint Restore (Dependent on TableName)
 
4. Add TableName and TableNameIndexHints test case

5. Add the ListFilterTrim utility function
pingcap/tidb#8532

### CODE REVIEW

File: ast/dml.go [hunk 1]
```diff
@@ -166,6 +229,29 @@ type IndexHint struct {
...
+		tempIndexNames = append(tempIndexNames, value.String())
+	}
+	indexNames := "(" + strings.Join(tempIndexNames, ",") + ")"
+
+	result := []string{
+		indexHintType,
+		indexHintScope,
+		indexNames,
+	}
+	sb.WriteString(strings.Join(ListFilterTrim(result), " "))
```
Comment (kennytm) [inline ast/dml.go]: You could just build the `sb` incrementally

```go
sb.WriteString(indexHintType)
if len(indexHintScope) > 0 {
    sb.WriteString(" ")
    sb.WriteString(indexHintScope)
}
sb.WriteString(" (")
for i, value := range n.IndexNames {
    if i > 0 {
        sb.WriteString(", ")
    }
    WriteName(sb, value.O)
}
sb.WriteString(")")
```

File: ast/dml_test.go [hunk 2]
```diff
@@ -72,3 +74,117 @@ func (ts *testDMLSuite) TestDMLVisitorCover(c *C) {
...
+	return []tableNameTestCase{
+		{"select * from t use index (hello)", "SELECT * FROM `t` USE INDEX (hello)"},
+		{"select * from t use index (hello, world)", "SELECT * FROM `t` USE INDEX (hello,world)"},
+		{"select * from t use index ()", "SELECT * FROM `t` USE INDEX ()"},
+		{"select * from t use key ()", "SELECT * FROM `t` USE INDEX ()"},
+		{"select * from t ignore key ()", "SELECT * FROM `t` IGNORE INDEX ()"},
+		{"select * from t force key ()", "SELECT * FROM `t` FORCE INDEX ()"},
+		{"select * from t use index for order by (idx1)", "SELECT * FROM `t` USE INDEX FOR ORDER BY (idx1)"},
+
+		{"select * from t use index (hello, world, yes) force key (good)", "SELECT * FROM `t` USE INDEX (hello,world,yes) FORCE INDEX (good)"},
```
Comment (kennytm) [inline ast/dml_test.go]: Please add a test for something like

```sql
SELECT * FROM t USE INDEX (`foo``bar`) FORCE INDEX (`baz``1`, `"xyz"`)
```

File: ast/dml.go (hunk 1 above)
Comment (feloxx) [reply ast/dml.go]: thank you, I think it is complicated, I will modify it now.

File: ast/dml.go [hunk 3]
```diff
@@ -166,6 +229,29 @@ type IndexHint struct {
 	HintScope  IndexHintScope
 }
 
+func (n *IndexHint) Restore(sb *strings.Builder) error {
+
+	indexHintType := indexHintTypeMap[int(n.HintType)]
+
+	indexHintScope := indexHintScopeMap[int(n.HintScope)]
```
[the remark this answers is not in the document]
Comment (feloxx) [reply ast/dml.go]: Yes, I think it is complicated, I have modified it.

File: ast/dml_test.go (hunk 2 above)
Comment (feloxx) [reply ast/dml_test.go]: Has been added

File: ast/util.go [hunk 4]
```diff
@@ -72,4 +72,4 @@ func WriteName(sb *strings.Builder, name string) {
 // EscapeName escape the `name`
 func EscapeName(name string) string {
 	return strings.Replace(name, "`", "``", -1)
-}
+}
```
Comment (kennytm) [inline ast/util.go]: Add back the trailing new line 😄

File: ast/dml.go [hunk 5]
```diff
@@ -135,7 +135,53 @@ type TableName struct {
...
+	//Joining indexHints.
+	indexHints := func(sb *strings.Builder) error {
+
+		for i, value := range n.IndexHints {
+
+			if i > 0 {
+				sb.WriteString(" ")
+			}
+			if err := value.Restore(sb); err != nil {
+				return errors.New("An error occurred while splicing IndexHints")
```
[the remark this answers is not in the document]
Comment (feloxx) [reply ast/dml.go]: Thank you for your advice.

File: ast/dml_test.go [hunk 6]
```diff
@@ -72,3 +74,128 @@ func (ts *testDMLSuite) TestDMLVisitorCover(c *C) {
...
+		{"select * from t ignore index (`foo``bar`) ignore key for order by (`baz``1`, xyz)","SELECT * FROM `t` IGNORE INDEX (`foo``bar`) IGNORE INDEX FOR ORDER BY (`baz``1`, `xyz`)"},
+
+		{"select * from t use index for group by (`foo``bar`) use index for order by (`baz``1`, `xyz`)","SELECT * FROM `t` USE INDEX FOR GROUP BY (`foo``bar`) USE INDEX FOR ORDER BY (`baz``1`, `xyz`)"},
+		{"select * from t use index for group by (`foo``bar`) ignore key for order by (`baz``1`, `xyz`)","SELECT * FROM `t` USE INDEX FOR GROUP BY (`foo``bar`) IGNORE INDEX FOR ORDER BY (`baz``1`, `xyz`)"},
+		{"select * from t use index for group by (`foo``bar`) ignore key for group by (`baz``1`, `xyz`)","SELECT * FROM `t` USE INDEX FOR GROUP BY (`foo``bar`) IGNORE INDEX FOR GROUP BY (`baz``1`, `xyz`)"},
+		{"select * from t use index for order by (`foo``bar`) ignore key for group by (`baz``1`, `xyz`)","SELECT * FROM `t` USE INDEX FOR ORDER BY (`foo``bar`) IGNORE INDEX FOR GROUP BY (`baz``1`, `xyz`)"},
+	}
+
+}
+func (tc *testDMLSuite) TestTableNameIndexHintsResotre(c *C) {
```
[the remark this answers is not in the document]
Comment (feloxx) [reply ast/dml_test.go]: sorry, I am careless.

File: ast/dml.go [hunk 7]
```diff
@@ -135,7 +135,53 @@ type TableName struct {
...
+		sb.WriteString("")
+	case n != nil && n.text != "":
+		WriteName(sb, n.text)
+	default:
+		if n.Schema.String() != "" {
+			WriteName(sb, n.Schema.String())
+			sb.WriteString(".")
+			WriteName(sb, n.Name.String())
+
+			sb.WriteString(" ")
```
Comment (leoppro) [inline ast/dml.go]: could you reuse this lines(L167-L172)?

File: ast/dml.go [hunk 8]
```diff
@@ -135,7 +135,53 @@ type TableName struct {
...
+		return nil
+
+	}
+
+	//Other logic may be empty, so you need to judge here.
+	switch {
+	case n == nil:
+		sb.WriteString("")
+	case n != nil && n.text != "":
+		WriteName(sb, n.text)
```
[the remark this answers is not in the document]
Comment (leoppro) [reply ast/dml.go]: and I don't think we need to check if n is nil.

File: ast/dml_test.go [hunk 9]
```diff
@@ -15,7 +15,9 @@ package ast_test
 
 import (
 	. "github.com/pingcap/check"
+	"github.com/pingcap/parser"
 	. "github.com/pingcap/parser/ast"
+	"strings"
```
Comment (leoppro) [inline ast/dml_test.go]:
```suggestion
	"strings"
	
	. "github.com/pingcap/check"
	"github.com/pingcap/parser"
	. "github.com/pingcap/parser/ast"
```

We separate the interior imports and exterior imports with blank lines

File: ast/dml.go [hunk 10]
```diff
@@ -166,6 +212,48 @@ type IndexHint struct {
 	HintScope  IndexHintScope
 }
 
+func (n *IndexHint) Restore(sb *strings.Builder) error {
+
+	indexHintType := ""
+	switch n.HintType {
+	case 1:
```
[the remark this answers is not in the document]
Comment (feloxx) [reply ast/dml.go]: 1.HintType and HintScope, only indexHint is used, so I don't wrap the `Restore` method.
2. add

File: ast/dml.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (feloxx) [reply ast/dml.go]: There is a ReferTable member in CreateTableStmt. 

If the statement does not have the ReferTable logic, there will be a null pointer problem. 

It is recommended to make a null judgment when the child member is `Restored`.

[the remark this answers is not in the document]
Comment (leoppro) [reply ast/dml.go]: i support that check it in CreateTableStmt

File: ast/dml.go [hunk 11]
```diff
@@ -166,6 +202,57 @@ type IndexHint struct {
...
+		indexHintType = "USE INDEX"
+	case 2:
+		indexHintType = "IGNORE INDEX"
+	case 3:
+		indexHintType = "FORCE INDEX"
+	default: // Prevent accidents
+		indexHintType = ""
```
Comment (leoppro) [inline ast/dml.go]: i support return error here

File: ast/dml.go [hunk 12]
```diff
@@ -135,7 +135,43 @@ type TableName struct {
...
+				return errors.Annotate(err, "An error occurred while splicing IndexHints")
+			}
+
+		}
+
+		return nil
+
+	}
+
+	if n.Schema.String() != "" {
```
Comment (leoppro) [inline ast/dml.go]:
```suggestion
	if n.Schema.String() != "" {
		WriteName(sb, n.Schema.String())
		sb.WriteString(".")
	}
	WriteName(sb, n.Name.String())
 	if err := indexHints(sb); err != nil {
		return errors.Annotate(err, "An error occurred while splicing TableName")
	}
```

File: ast/dml.go [hunk 13]
```diff
@@ -166,6 +202,57 @@ type IndexHint struct {
...
+	case 1:
+		indexHintScope = ""
+	case 2:
+		indexHintScope = "FOR JOIN"
+	case 3:
+		indexHintScope = "FOR ORDER BY"
+	case 4:
+		indexHintScope = "FOR GROUP BY"
+	default: // Prevent accidents
+		indexHintType = ""
```
Comment (leoppro) [inline ast/dml.go]: ditto

File: ast/dml.go [hunk 14]
```diff
@@ -166,6 +202,57 @@ type IndexHint struct {
...
+	}
+
+	if len(indexHintType) > 0 {
```
Comment (leoppro) [inline ast/dml.go]: we needn't check here.

File: ast/dml.go [hunk 15]
```diff
@@ -166,6 +202,57 @@ type IndexHint struct {
...
+		sb.WriteString(indexHintType)
+	}
+
+	if len(indexHintScope) > 0 {
```
Comment (leoppro) [inline ast/dml.go]: maybe we don;t check here too,and add space to indexHintScope:
```
switch n.HintScope {
	case 1:
		indexHintScope = ""
	case 2:
		indexHintScope = " FOR JOIN"
	case 3:
		indexHintScope = " FOR ORDER BY"
	case 4:
		indexHintScope = " FOR GROUP BY"
	...
```

File: ast/dml.go (hunk 11 above)
Comment (feloxx) [reply ast/dml.go]: Add default return new error

File: ast/dml_test.go [hunk 16]
```diff
@@ -72,3 +72,67 @@ func (ts *testDMLSuite) TestDMLVisitorCover(c *C) {
 		v.node.Accept(visitor1{})
 	}
 }
+
+type tableNameTestCase struct {
```
Comment (leoppro) [inline ast/dml_test.go]: remove unused `tableNameTestCase`

File: ast/dml_test.go [hunk 17]
```diff
@@ -72,3 +72,67 @@ func (ts *testDMLSuite) TestDMLVisitorCover(c *C) {
...
+	sourceSQL string
+	expectSQL string
+}
+
+/* ***********************************************************************************************************
+ * TableName Stmt Test Case
+ */
+// only TableName test data
```
Comment (leoppro) [inline ast/dml_test.go]: remove this 
```
/* ***********************************************************************************************************
 * TableName Stmt Test Case
 */
// only TableName test data
```

File: ast/dml.go [hunk 18]
```diff
@@ -166,6 +188,51 @@ type IndexHint struct {
...
+	case 3:
+		indexHintScope = " FOR ORDER BY"
+	case 4:
+		indexHintScope = " FOR GROUP BY"
+	default: // Prevent accidents
+		return errors.New("IndexHintScope has an error while matching")
+	}
+
+	sb.WriteString(indexHintType)
+	if len(indexHintScope) > 0 {
```
Comment (leoppro) [inline ast/dml.go]: we needn't `if len(indexHintScope) > 0 {`

**Comment by leoppro** [comment]
LGTM, PTAL again @kennytm

File: ast/dml.go [hunk 19]
```diff
@@ -135,7 +135,29 @@ type TableName struct {
 
 // Restore implements Recoverable interface.
 func (n *TableName) Restore(sb *strings.Builder) error {
-	return errors.New("Not implemented")
+	//Joining indexHints.
+	indexHints := func(sb *strings.Builder) error {
+		for i, value := range n.IndexHints {
+			if i >= 0 {
```
Comment (kennytm) [inline ast/dml.go]: Remove the condition, `i >= 0` is always true.

File: ast/dml.go [hunk 20]
```diff
@@ -135,7 +135,29 @@ type TableName struct {
...
+		}
+		return nil
+	}
+
+	if n.Schema.String() != "" {
+		WriteName(sb, n.Schema.String())
+		sb.WriteString(".")
+	}
+	WriteName(sb, n.Name.String())
+	if err := indexHints(sb); err != nil {
```
Comment (kennytm) [inline ast/dml.go]: Since `indexHints` is only used once here, perhaps just inline that function directly?

File: ast/dml.go (hunk 19 above)
Comment (feloxx) [reply ast/dml.go]: thank you

**Pull request merged**